OpenRA.Mods.Common/Effects/Missile.cs

```diff
@@ -141,52 +212,138 @@ public void Tick(World world)
 			if (anim != null)
 				anim.Tick();
 
-			// Missile tracks target
+			var cell = world.Map.CellContaining(pos);
+			terrainHeight = world.Map.MapHeight.Value[cell] * 512;
+
+			// Switch from freefall mode to homing mode
+			if (ticks == info.ActivationDelay + 1)
+			{
+				activated = true;
+				hFacing = OpenRA.Traits.Util.GetFacing(velocity, hFacing);
+				speed = velocity.Length;
+			}
+
+			// Switch from homing mode to freefall mode
+			if (info.RangeLimit != 0 && ticks == info.RangeLimit + 1)
+			{
+				activated = false;
+				velocity = new WVec(0, -speed, 0)
+					.Rotate(new WRot(WAngle.FromFacing(vFacing), WAngle.Zero, WAngle.Zero))
+					.Rotate(new WRot(WAngle.Zero, WAngle.Zero, WAngle.FromFacing(hFacing)));
+			}
+
+			// Check if target position should be updated (actor visible & locked on)
 			if (args.GuidedTarget.IsValidFor(args.SourceActor) && lockOn)
 				targetPosition = args.GuidedTarget.CenterPosition;
 
+			// Compute current distance from target position
 			var dist = targetPosition + offset - pos;
-			var desiredFacing = OpenRA.Traits.Util.GetFacing(dist, facing);
-			var desiredAltitude = targetPosition.Z;
-			var jammed = info.Jammable && world.ActorsWithTrait<JamsMissiles>().Any(JammedBy);
+			var len = dist.Length;
 
-			if (jammed)
+			WVec move;
+			if (activated)
+			{
+				// If target is within range, keep speed constant and aim for the target.
+				// The speed needs to be kept constant to keep range computation relatively simple.
+
+				// If target is not within range, accelerate the projectile. If cruise altitudes
+				// are not used, aim for the target. If the cruise altitudes are used, aim for the
+				// target horizontally and for cruise altitude vertically.
+
+				// Target is considered in range if after an additional tick of accelerated motion
+				// the horizontal distance from the target would be less than
+				// the diameter of the circle that the missile travels along when
+				// turning vertically at the maximum possible rate.
+				// This should work because in the worst case, the missile will have to make
+				// a semi-loop before hitting the target.
+				var hLenCurr = dist.HorizontalLength; // Get current distance from target
+				var hLenNext = (long)(hLenCurr - speed - info.Acceleration.Length).Clamp(0, hLenCurr); // Get distance from target in next tick
+				bool inRange = hLenNext * hLenNext * info.VerticalRateOfTurn * info.VerticalRateOfTurn * 3.14 * 3.14
```

Comment:
This condition might change, but @GraionDilach, right now it works well with any firing angle.